File: questao_15/questao_15.java
class SaldoInsuficienteException extends Exception {
    public SaldoInsuficienteException(String mensagem) {
        super(mensagem);
    }
}

class ContaBancaria {
    @SuppressWarnings("unused")
    private String titular;
    private double saldo;

    public ContaBancaria(String titular, double saldo) {
        this.titular = titular;
        this.saldo = saldo;
    }

    public void depositar(double valor) {
        saldo += valor;
        System.out.printf("Deposito de R$ %.2f realizado. Saldo atual: R$ %.2f%n", valor, saldo);
    }

    public void sacar(double valor) throws SaldoInsuficienteException {
        if (valor > saldo) {
            throw new SaldoInsuficienteException("Saldo insuficiente para realizar o saque.");
        }
        saldo -= valor;
        System.out.printf("Saque de R$ %.2f realizado. Saldo atual: R$ %.2f%n", valor, saldo);
    }

    public static void main(String[] args) {
        ContaBancaria conta = new ContaBancaria("Natan", 180);

        try {
            conta.sacar(250);
        } catch (SaldoInsuficienteException e) {
            System.out.println(e.getMessage());
        }
    }
}

File: questao_15/questao_15.py
# Exceções/Erros Personalizados: Crie uma classe de exceção personalizada SaldoInsuficienteException em Java e Python, ou error em Golang,
# que seja lançada quando houver uma tentativa de saque superior ao saldo disponível na classe ContaBancaria

class SaldoInsuficienteException(Exception):
    def __init__(self, mensagem):
        super().__init__(mensagem)

class ContaBancaria:
    def __init__(self, titular, saldo=0):
        self.titular = titular
        self.saldo = saldo

    def depositar(self, valor):
        self.saldo += valor
        print(f"Deposito de R$ {valor:.2f} realizado. Saldo atual: R$ {self.saldo:.2f}")

    def sacar(self, valor):
        if valor > self.saldo:
            raise SaldoInsuficienteException("Saldo insuficiente para realizar o saque.")
        self.saldo -= valor
        print(f"Saque de R$ {valor:.2f} realizado. Saldo atual: R$ {self.saldo:.2f}")

conta = ContaBancaria("Natan", 180)

try:
    conta.sacar(250)
except SaldoInsuficienteException as e:
    print(e)
